### PR TITLE
[pipeline] add bert_for_pretraining bert_lmhead forward and policy

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4119 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR adds bert_for_pretraining bert_lmhead forward and policy and the responding tests,  fix the forward of base Bert

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/policy/bert.py [hunk 1]
```diff
@@ -285,51 +285,76 @@ def bert_for_pretraining_forward(
...
-
-    return BertForPreTrainingOutput(
-        loss=total_loss,
-        prediction_logits=prediction_scores,
-        seq_relationship_logits=seq_relationship_score,
-        hidden_states=outputs.hidden_states,
-        attentions=outputs.attentions,
-    )
+    else:
+        if not return_dict:
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:344]: Intermediate layers should always return a dict.

File: colossalai/pipeline/policy/bert.py [hunk 2]
```diff
@@ -285,51 +285,76 @@ def bert_for_pretraining_forward(
...
+                all_hidden_states,
+                all_self_attentions,
+                all_cross_attentions,
+            ] if v is not None)
 
 
 class BertForPreTrainingPolicy(Policy):
 
     def __init__(self, stage_manager: PipelineStageManager, num_layers: int, num_stages: int):
+        super().__init__(stage_manager=stage_manager)
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:357]: Num_stages can be got from stage manager

Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py:357]: done

File: colossalai/pipeline/policy/bert.py [hunk 3]
```diff
@@ -285,53 +305,74 @@ def bert_for_pretraining_forward(
...
-    return BertForPreTrainingOutput(
-        loss=total_loss,
-        prediction_logits=prediction_scores,
-        seq_relationship_logits=seq_relationship_score,
-        hidden_states=outputs.hidden_states,
-        attentions=outputs.attentions,
-    )
+    else:
+        if not return_dict:
+            logger.warning_once('The output of intermediate stage should always be a dict')
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:366]: Warning is not needed, since intermdiate outputs are transparent to users.

Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

File: colossalai/pipeline/policy/bert.py [hunk 4]
```diff
@@ -221,34 +241,33 @@ def custom_forward(*inputs):
...
-        last_hidden_state=hidden_states,
-        past_key_values=next_decoder_cache,
-        hidden_states=all_hidden_states,
-        attentions=all_self_attentions,
-        cross_attentions=all_cross_attentions,
-    )
+    # output of non-first and non-last stages: must be a dict
+    else:
+        # intermediate stage always return dict
+        return BertModelIntermediateOutput(hidden_states=hidden_states,)
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:257]: `BertModelIntermediateOutput` has compatibility problem with pytree, using original `dict` is better.

File: colossalai/pipeline/policy/bert.py [hunk 5]
```diff
@@ -285,53 +304,73 @@ def bert_for_pretraining_forward(
...
-    return BertForPreTrainingOutput(
-        loss=total_loss,
-        prediction_logits=prediction_scores,
-        seq_relationship_logits=seq_relationship_score,
-        hidden_states=outputs.hidden_states,
-        attentions=outputs.attentions,
-    )
+    else:
+        # intermediate stage always return dict
+        return BertModelIntermediateOutput(hidden_states=hidden_states,)
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:365]: Same problem

File: colossalai/pipeline/policy/bert.py (hunk 4 above)
Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

File: colossalai/pipeline/policy/bert.py (hunk 5 above)
Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

File: colossalai/pipeline/policy/bert.py [hunk 6]
```diff
@@ -355,22 +383,141 @@ def get_shared_params(self, module: BertForPreTraining) -> List[Dict[int, Tensor
...
+        hold_layers.extend(module.bert.encoder.layer[start_idx:end_idx])
+        if self.stage_manager.is_last_stage():
+            hold_layers.append(module.bert.pooler)
+            hold_layers.append(module.cls)
+
+        return hold_layers
+
+    def get_shared_params(self, module: BertLMHeadModel) -> List[Dict[int, Tensor]]:
+        '''no shared params in bertmodel'''
+        pass
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:520]: If no sharded params, it should return an empty list.

Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

**Pull request merged**